crates/ra_editor/src/diagnostics.rs

```diff
@@ -0,0 +1,269 @@
+        match location {
+            Location::Offset(offset) => TextRange::offset_len(offset, 1.into()),
+            Location::Range(range) => range,
+        }
+    }
+
+    let mut errors: Vec<Diagnostic> = file
+        .errors()
+        .into_iter()
+        .map(|err| Diagnostic {
+            range: location_to_range(err.location()),
+            msg: format!("Syntax Error: {}", err),
+            severity: Severity::Error,
+            fix: None,
+        })
+        .collect();
+
+    errors.extend(check_unnecessary_braces_in_use_statement(file));
+    errors.extend(check_struct_shorthand_initialization(file));
+    errors
+}
+
+fn check_unnecessary_braces_in_use_statement(file: &SourceFileNode) -> Vec<Diagnostic> {
+    let mut diagnostics = Vec::new();
+    for node in file.syntax().descendants() {
+        if let Some(use_tree_list) = ast::UseTreeList::cast(node) {
+            if let Some((single_use_tree,)) = use_tree_list.use_trees().collect_tuple() {
+                let range = use_tree_list.syntax().range();
+                let edit = text_edit_for_remove_unnecessary_braces_with_self_in_use_statement(
+                    single_use_tree,
+                )
+                .unwrap_or_else(|| {
+                    let to_replace = single_use_tree.syntax().text().to_string();
+                    let mut edit_builder = TextEditBuilder::new();
+                    edit_builder.delete(range);
+                    edit_builder.insert(range.start(), to_replace);
+                    edit_builder.finish()
+                });
+
+                diagnostics.push(Diagnostic {
+                    range,
+                    msg: format!("Unnecessary braces in use statement"),
+                    severity: Severity::WeakWarning,
+                    fix: Some(LocalEdit {
+                        label: "Remove unnecessary braces".to_string(),
+                        edit,
+                        cursor_position: None,
+                    }),
+                })
+            }
+        }
+    }
+
+    diagnostics
+}
+
+fn text_edit_for_remove_unnecessary_braces_with_self_in_use_statement(
+    single_use_tree: ast::UseTree,
+) -> Option<TextEdit> {
+    let use_tree_list_node = single_use_tree.syntax().parent()?;
+    if single_use_tree
+        .path()?
+        .segment()?
+        .syntax()
+        .first_child()?
+        .kind()
+        == SyntaxKind::SELF_KW
+    {
+        let start = use_tree_list_node.prev_sibling()?.range().start();
+        let end = use_tree_list_node.range().end();
+        let range = TextRange::from_to(start, end);
+        let mut edit_builder = TextEditBuilder::new();
+        edit_builder.delete(range);
+        return Some(edit_builder.finish());
+    }
+    None
+}
+
+fn check_struct_shorthand_initialization(file: &SourceFileNode) -> Vec<Diagnostic> {
+    let mut diagnostics = Vec::new();
+    for node in file.syntax().descendants() {
```

Comment:
One move to simplify this would be

`for named_field in file.syntax().descendats.filter_map(ast::NamedField::cast)`